### PR TITLE
misc: enhance support for replayable instances of `InputStream`

### PR DESCRIPTION
**Pull request by lauzadis**
## Description of changes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/content/ByteStreamJVM.kt [hunk 1]
```diff
@@ -114,11 +114,25 @@ public fun ByteStream.Companion.fromInputStream(
...
+                return object : SdkSource by source() {
+                    // no-op close
+                    override fun close() { }
+                }
+            }
+
+            return source()
+        }
     }
 }
```
Comment (ianbotsf) [inline runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/content/ByteStreamJVM.kt:137]: **Question**: Is this a behavioral change wrt closing? It seems like before we returned the raw underlying `SdkSource` before which, if closed, would close the underlying `InputStream`. Now it looks like that won't happen...why is that necessary?

Comment (lauzadis) [reply runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/content/ByteStreamJVM.kt:137]: Yes it is. Without this change the underlying `InputStream` is closed after hashing (because we call a `.use { }` on the `SdkSource`) and subsequent attempts to read the body (e.g. during transmission) fail with `IOException: Stream closed`

**Review by 0marperez — APPROVED** [review]
